### PR TITLE
Issue #13321: Kill mutation for DescendantTokenCheck

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13321: Kill mutation for DescendantTokenCheck

----------

# Check 
https://checkstyle.org/checks/misc/descendanttoken.html#DescendantToken

---------

# mutation 
https://github.com/checkstyle/checkstyle/blob/20a52f70548476a4ca3bb0f56e4d8fcd9cd775b7/config/pitest-suppressions/pitest-misc-suppressions.xml#L3-L19

-----------

# Explaination

Test added

----------

# Regression

Diff Regression config: https://gist.githubusercontent.com/Kevin222004/14aa273cb63f5e0c9b09b80021c9662d/raw/e44a7eed1601ecce19ab969e66dd27c698281ebd/desc.xml
Diff Regression projects: https://gist.githubusercontent.com/Kevin222004/21e3934e85f802e2fbd48af06d122364/raw/604256badd733d8568064f371d55657c04b00dfd/test-projects-2.properties
Report label: Regression-1

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
@Kevin222004  please generate reports and fix CI failures

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by romani** [comment]
@Kevin222004 , please continue.

**Comment by romani** [comment]
conflict with spotbugs:
```
INFO] --- spotbugs-maven-plugin:4.7.3.0:check (default) @ checkstyle ---
[INFO] BugInstance size is 5
[INFO] Error size is 0
[INFO] Total bugs: 5
[ERROR] Low: DescendantTokenCheck.counts not initialized in constructor and dereferenced in com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck.countTokens(DetailAST, int) [com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck] At DescendantTokenCheck.java:[line 787] UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR
[ERROR] Low: DescendantTokenCheck.counts not initialized in constructor and dereferenced in com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck.logAsSeparated(DetailAST) [com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck] At DescendantTokenCheck.java:[line 713] UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR
[ERROR] Low: DescendantTokenCheck.counts not initialized in constructor and dereferenced in com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck.logAsTotal(DetailAST) [com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck] At DescendantTokenCheck.java:[line 754] UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR
[ERROR] Low: DescendantTokenCheck.limitedTokens not initialized in constructor and dereferenced in com
.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck.logAsSeparated(DetailAST) [com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck] At DescendantTokenCheck.java:[line 712] UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR
[ERROR] Low: DescendantTokenCheck.limitedTokens not initialized in constructor and dereferenced in com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck.logAsTotal(DetailAST) [com.puppycrawl.tools.checkstyle.checks.DescendantTokenCheck] At DescendantTokenCheck.java:[line 753] UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR

```

**Comment by romani** [comment]
we set `counts` at
 https://github.com/checkstyle/checkstyle/blob/e539d5113a311f2e60098f3c9d2872aad07c665c/src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java#L806-L817

we need to suppress spotbugs violations. with comment  `spotbugs does not know sequence of public method executions, null dereference is not possible` But it is not a problem.

**Comment by romani** [comment]
@Kevin222004 , please do excludes for spotbugs at https://github.com/checkstyle/checkstyle/blob/master/config/spotbugs-exclude.xml

**Comment by romani** [comment]
@Kevin222004 , please fix Checker

**Comment by Kevin222004** [comment]
@romani I will fix the checker

but i dont know how can i fix https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=15736&view=logs&j=c902ebb4-c9f8-5f09-4e17-ff78fbbc842e&t=9ca98c81-ff64-58f0-9d03-a23ac1c4a111&l=1123

**Comment by romani** [comment]
Add to suppression it for now

**Comment by romani** [comment]
@Kevin222004 , this looks like simple PR to update to get approval.

**Comment by Kevin222004** [comment]
@romani I am sorry, I have given the time but I am not sure where and how I can suppress this.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java [hunk 1]
```diff
@@ -659,7 +659,7 @@ public class DescendantTokenCheck extends AbstractCheck {
     private boolean sumTokenCounts;
     /** Specify set of tokens with limited occurrences as descendants. */
     @XdocsPropertyType(PropertyType.TOKEN_ARRAY)
-    private int[] limitedTokens = CommonUtil.EMPTY_INT_ARRAY;
+    private int[] limitedTokens;
```
Comment (Vyom-Yadav) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java:662]: @Kevin222004 Won't this throw NPE when there is no mention of this property in the config? `limitedTokens` is just not there in the config. In that case `counts` will be null and `Arrays.fill(counts, 0);` will throw an NPE.

**Review by Vyom-Yadav — COMMENTED** [review]
Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java (hunk 1 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java:662]:
```diff
/*
DescendantToken
tokens = LITERAL_SWITCH
maximumDepth = 2
- limitedTokens =
minimumNumber = 1

*/
```
In this type of config, an NPE would be thrown. Is `limitedTokens` mandatory for the user to set?

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java]: I am not sure where things have gone wrong from my side. but done

Test Added

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java [hunk 2]
```diff
@@ -276,4 +276,11 @@ public void testMaxTokenTypeReverseOrder() throws Exception {
                 getPath("InputDescendantTokenLastTokenType2.java"), expected);
     }
 
+    @Test
+    public void testProperty() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
```
Comment (Vyom-Yadav) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java:281]: `testProperty` is really vague, please keep a specific name.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/descendanttoken/InputDescendantTokenProperty.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+package com.puppycrawl.tools.checkstyle.checks.descendanttoken;
+
+// ok
+public class InputDescendantTokenProperty {
+}
+
+class Test {
+    public static void main(String[] args) {
+        int x = 1;
+        switch (x) { // ok
```
Comment (Vyom-Yadav) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/descendanttoken/InputDescendantTokenProperty.java:18]: One `ok` on top is enough.

**Review by Vyom-Yadav — CHANGES_REQUESTED** [review]
minors:

**Review by romani — APPROVED** [review]
Ok to merge as raised items are resolved

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java]: @Kevin222004  , please fix method name

Comment (Kevin222004) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/descendanttoken/InputDescendantTokenProperty.java (hunk 3 above)
Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/descendanttoken/InputDescendantTokenProperty.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java]: Input file has values that are not default,

Comment (Kevin222004) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheckTest.java]: I have make it simple testLimitedToken

**Review by Vyom-Yadav — APPROVED** [review]
LGTM!

**Comment by Vyom-Yadav** [comment]
@romani Assigned it to you as 2 reviews are sufficient for PRs adding a test case to kill a mutation.

**Pull request merged**